### DISCUSSION TITLE
Dragging objects in playfield should not change its time in DS mode

### DISCUSSION THREAD

**Discussion by cihe13375** [ppy/osu#23815] [Ideas] [answered]
Currently under distance snap mode, one may change the time of an object by dragging it in the playfield:
![2023-06-08 17 52 39](https://github.com/ppy/osu/assets/33959010/f750814a-319d-444e-8eab-880f31e7b185)
![2023-06-08 17 52 55](https://github.com/ppy/osu/assets/33959010/1eca4b2c-b145-407a-9330-79ab691984df)
This feature kinda annoys me, because it's too prone to dragging the object to a wrong divisor (e.g. accidentally change a 1/4 gap to a 1/2 gap), especially when e.g. making variable spacing streams. Furthermore, it makes low DS (like 0.2x) almost unusable because the grids are so dense that it's difficult to keep the objected snapped to the correct grid.
Generally speaking, I think this feature doesn't make much sense, since one determines the time of an object base on rhythm, not on its location in the playfield. In other words, the time is determined before the object is placed down (like the usual way of placing objects), and moving it in the playfield should not change that.

**Comment by bdach** [comment]
> since one determines the time of an object base on rhythm, not on its location in the playfield

Isn't the entire point of the grid to make sure that the spacing between circles matches the timing of the circles, so that the player can move their mouse at a constant speed and still hit the notes? Allowing moving the notes without changing timing breaks this. It also makes the colouring of the distance snap grid no longer make sense.

I get that you may want to purposefully break this convention, but at that point do you even *need* the distance snap grid?

**Reply by cihe13375** [reply]
> Allowing moving the notes without changing timing breaks this

When I drag an object in playfield in DS mode, I expect it to move only on that *specific* concentric circle, not jumping to others. The distance is not changed at all (so DS is always kept), only the direction is variable. This is what currently stable does. The problem of current lazer behavior is that one may accidently drag the object to another beat divisor, resulting in wrongly snapped objects.

**Reply by cihe13375** [reply]
The logic of DS, as I believe, is that the mapper first selects the rhythm (by seeking to a specific position in timeline), and the distance automatically follows the time interval. Currently lazer goes the inverse (the mapper selects a distance in the playfield, and the time of object follows that distance).

**Comment by peppy** [comment]
Distance snap is a time to distance mapping. The behaviour is correct.

Yes, it behaves differently to stable, but once you adjust it's much more ergonomic.

If you prefer the old behaviour, aim for the brightest line, which is your current point in time.

**Reply by cihe13375** [reply]
Alright, I guess I'll just switch back to stable for DS heavy mapping...

**Reply by peppy** [reply]
Maybe we can add a hotkey to hold to lock to the current ring?

**Reply by peppy** [reply, marked as answer]
https://github.com/ppy/osu/pull/23837

**Reply by cihe13375** [reply]
I appreciate the update and your hard work, but it doesn't really solve the original problem. Indeed with the "limit to current time" setting one may accidentally change the time of placed objects in more surprising ways (stable doesn't behave like this, of course):
initial state:
![before](https://github.com/ppy/osu/assets/33959010/c6241c8d-073f-4a8a-ba17-35fa2a844b8a)
after dragging circle 4 in DS mode:
![after](https://github.com/ppy/osu/assets/33959010/bbeeeac6-0e36-44c4-a9d1-c8791c65bfbb)

**Reply by peppy** [reply]
Please make a video. I can't tell what's going on from both your text and the images.

**Reply by cihe13375** [reply]
stable:

https://github.com/ppy/osu/assets/33959010/36d8af0e-2af1-42d6-aa5d-7f390ea16bc9

lazer (default):

https://github.com/ppy/osu/assets/33959010/e840be6a-a709-4b71-9a08-12971b786a2d

lazer (limit to current time):

https://github.com/ppy/osu/assets/33959010/a366b2e2-5e09-4b21-9906-8b29a5ba180f

**Reply by peppy** [reply]
Okay that's quite fringe behaviour. I'm not sure what makes most sense here.

To summarise in words:

When in select mode with distance snap on and adjusting a future hitobject, in lazer it will possibly bring the selected object in front of other objects it was previous ahead of. In stable that won't happen.
